### PR TITLE
util/types: new Compare implementation with Datum.

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: util/types/datum.go [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+func parseFloatLoose(s string) float64 {
+	s = numPart(s)
+	val, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		log.Errorf("should not happen!")
+	}
+	return val
+}
+
+func (d *Datum) compareString(s string) int {
```
Comment (shenli) [inline util/types/datum.go]: Miss collation info.

Comment (coocood) [inline util/types/datum.go]: The old implementation doesn't have collation info.

Comment (coocood) [inline util/types/datum.go]: It's easy to add collation info in the future if we need to support collation compare..

File: executor/subquery.go [hunk 2]
```diff
@@ -26,7 +26,7 @@ var _ ast.SubqueryExec = &subquery{}
 
 // SubQuery is an exprNode with a plan.
```
Comment (shenli) [inline executor/subquery.go]: SubQuery -> subquery

Comment (coocood) [inline executor/subquery.go]: Will fix in following PR.

File: util/types/datum.go [hunk 3]
```diff
@@ -0,0 +1,519 @@
...
+	KindMysqlEnum
+	KindMysqlHex
+	KindMysqlSet
+	KindMysqlTime
+	KindRow
+	KindInterface
+)
+
+// Datum is a data box holds different kind of data.
+// It has better performance and easier to use than `interface{}`.
```
Comment (shenli) [inline util/types/datum.go]: is easier

File: util/types/datum.go [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+	"strings"
+
+	"github.com/ngaut/log"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/hack"
+)
+
+// Kind constants.
+const (
+	KindNull  int = 0
```
Comment (shenli) [inline util/types/datum.go]: KindNull int = iota

Comment (coocood) [inline util/types/datum.go]: This is more explicit. So we can easily tell `KindNull` is 0.

File: util/types/datum.go [hunk 5]
```diff
@@ -0,0 +1,519 @@
...
+	for i := 0; i < len(row) && i < len(dRow); i++ {
+		cmp := dRow[i].compareDatum(row[i])
+		if cmp != 0 {
+			return cmp
+		}
+	}
+	return CompareInt64(int64(len(dRow)), int64(len(row)))
+}
+
+// NewDatum create a new Datum from an interface{}.
```
Comment (shenli) [inline util/types/datum.go]: creates

File: util/types/datum.go [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+		return "0"
+	}
+	return s[:i]
+}
+
+func parseFloatLoose(s string) float64 {
+	s = numPart(s)
+	val, err := strconv.ParseFloat(s, 64)
+	if err != nil {
+		log.Errorf("should not happen!")
```
Comment (shenli) [inline util/types/datum.go]: mysql> select 1.5 > "xxx";
+-------------+
| 1.5 > "xxx" |
+-------------+
|           1 |
+-------------+
1 row in set, 1 warning (0.01 sec)

**Comment by shenli** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**